source/h1_encoder.c

```diff
@@ -233,35 +271,40 @@ int aws_h1_encoder_message_init_from_response(
 
     /* head-end: "\r\n" */
     size_t head_end_len = 2;
-    size_t head_total_len = response_line_len;
-    err |= aws_add_size_checked(header_lines_len, head_total_len, &head_total_len);
-    err |= aws_add_size_checked(head_end_len, head_total_len, &head_total_len);
+    *head_total_len = response_line_len;
+    err |= aws_add_size_checked(header_lines_len, *head_total_len, head_total_len);
+    err |= aws_add_size_checked(head_end_len, *head_total_len, head_total_len);
     if (err) {
-        goto error;
+        return AWS_OP_ERR;
     }
+    return AWS_OP_SUCCESS;
+}
+
+int aws_h1_encoder_message_init_from_response(
+    struct aws_h1_encoder_message *message,
+    struct aws_allocator *allocator,
+    const struct aws_http_message *response,
+    bool body_headers_ignored) {
 
+    AWS_ZERO_STRUCT(*message);
+
+    message->body = aws_http_message_get_body_stream(response);
+    /**
+     * Calculate total size needed for outgoing_head_buffer, then write to buffer.
+     */
+    size_t head_total_len;
+    int err = s_get_response_header_len(message, &head_total_len, response, body_headers_ignored);
+    if (err) {
+        goto error;
+    }
     err = aws_byte_buf_init(&message->outgoing_head_buf, allocator, head_total_len);
     if (err) {
         return AWS_OP_ERR;
     }
-
-    bool wrote_all = true;
-
-    wrote_all &= aws_byte_buf_write_from_whole_cursor(&message->outgoing_head_buf, version);
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, ' ');
-    wrote_all &= aws_byte_buf_write_from_whole_cursor(&message->outgoing_head_buf, status_code);
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, ' ');
-    wrote_all &= aws_byte_buf_write_from_whole_cursor(&message->outgoing_head_buf, status_text);
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, '\r');
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, '\n');
-
-    s_write_headers(&message->outgoing_head_buf, response);
-
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, '\r');
-    wrote_all &= aws_byte_buf_write_u8(&message->outgoing_head_buf, '\n');
-    (void)wrote_all;
-    AWS_ASSERT(wrote_all);
-
+    err = s_write_response_headers_into_buffer(&message->outgoing_head_buf, response);
```

Comment:
trivial: was there a reason to have two separate functions? `s_get_response_header_len()` and `s_write_response_headers_into_buffer()`? Would it be simpler to have 1 function that took a pointer to the buffer it should operate on?